DIPs/DIP1NNN-CrazyPython.md

```diff
@@ -0,0 +1,283 @@
+outside it -->
+<!-- Todo: add unit tests -->
+<!-- Todo: write Kahn's Algorithm in D and showcase the various solutions -->
+
+Like normal D static arrays, dynamic-size static arrays are pass-by-value: 
+
+```dlang
+@nogc void no_op(int[?] list) {
+for (int i = 0; i < list.length; ++i) {
+list[i] = 0;
+}
+}
+unittest {
+int[5] array = [1, 2, 3, 4, 5];
+no_op(array);
+assert(array == [1, 2, 3, 4, 5]);
+}
+```
+
+A dynamic array can be implicitly converted to a dynamic-size static array as a
+copy:
+
+```
+unittest {
+int[] array = [1, 2, 3, 4, 5];
+no_op(array);
+assert(array == [1, 2, 3, 4, 5]);
+int[?] another_array = array;
+another_array[0] = 999;
+assert(array[0] != 999);
+}
+unittest {
+int[][] array = [[1, 2], [3, 4], [5, 6]];
+int[?][?] secondarray = array;
+}
+```
+
+Dynamic-size static arrays have a `.length` property. The `.length` property
+may be changed arbitrarily, just like for a dynamic array. 
+
+**Implementation Defined:** Enough space to hold the original length of the
+array is allocated on the stack.
+
+**Implementation Defined:** Shrinking a dynamic-size static array does not free
+or deallocate the memory held by a dynamic-size static array.
+
+**Implementation Defined:** Expanding a dynamic-sized array's `.length`
+property beyond its *original length* may allocate memory on the heap. 
+
+**Implementation Defined:** The compiler should avoid generating code for heap
+allocation when it can prove that a particular statement will not change the
+`.length` of the array beyond its original length. For example, `int[n] array;
+for (int i = 0; i < n; ++i) { array[i] = rand(); }` should never generate code
+that uses heap allocation. That's because the compiler can see that the array
+is never accessed beyond its original size. Similarly for `int[n] array;
+array.length = 0; while (array.length < n) { array ~= rand(); }`.
+
+**Undefined behavior:** Accessing a dynamic-size static array by index beyond
+its `.length`.
+
+
+A declaration of the form `T[?] identifier;` has an *original length* of
+zero. 
+
+`int[expr] a;`
+
+When `expr` is not always known at compile-time, declares a dynamic-size static
+array. `expr` is the *original length* of the dynamic-size static array. 
+
+```
+unittest {
+int[][] array = [[1, 2], [3, 4], [5, 6]];
+int[?][?] secondarray = array;
+}
+```
+
+Like static arrays, the memory of a dynamic-size static array is automatically
+deallocated when exiting the scope where they were allocated. 
+
+It is illegal to return a slice of a dynamic-size static array, because the
+dynamic-size static array would be deallocated on function return. 
+
+<!-- It is illegal to create a dynamic-size static array on the heap, for
+example with the `new` keyword. -->
+
+The `.length` property of a dynamic-size static array shall be taken into
+consideration by `foreach` and `static foreach`.
+
+If the `.length` property of a dynamic-sized static array is never set or read,
```

Comment:
I don't think this condition is actually possible to meaningfully implement, especially with metaprogramming and recursive functions.

Reply:
The compiler allocates memory for the `.length` property when it needs to know it at run-time. When the compiler does not need to know the size of the `.length` property at run-time, it does not allocate memory for it.

Example 1
```
void fun(int N)
{
  int[N] a;
  foreach (i; a)
  {
    writeln(i);
  }
}
```

The compiler does not need to allocate memory for the length of the array-- `int N` might as well be immutable, so it can just access that.

Example 2
```
void fun(int N)
{
  int[N] a;
  foreach (i; a)
  {
    writeln(i, ++N);
  }
}
```

The compiler needs to allocate memory for the length of the array, since N is changing.

---

I realize that it may not be economical in terms of compile-time to ensure that the `.sizeof` properties are the same regardless. In that case, my proposal's sizeof requirements probably need to be relaxed.